### ISSUE TITLE
Build s-2-i-dotnetcore-ex-git failed 

### ISSUE THREAD

**Issue by deenamanick** [redhat-developer/s2i-dotnetcore#314]
Hi, 

I am trying to test this repo in Openshift 4.3 Online. While building an dotnetcore image, it failed with the following error message. can you advise me how to fix this issue?

--
Cloning "https://github.com/redhat-developer/s2i-dotnetcore-ex.git" ...
	Commit:	1ded43a7b3737ac31e31eee352134fd7c8c0d424 (Update for 2.1 release.)
	Author:	Tom Deseyn <tom.deseyn@gmail.com>
	Date:	Mon May 28 10:08:30 2018 +0200
Caching blobs under "/var/cache/blobs".
Getting image source signatures
Copying blob sha256:305d73a95c8fece2b53a34e040df1c97eb6b7f7cc4e0a7933465f0b7325e3d72
Copying blob sha256:00f17e0b37b0515380a4aece3cb72086c0356fc780ef4526f75476bea36a2c8b
Copying blob sha256:995f8af182b62856ae62fcbff232dc3c966d4030b36cfd139a56f4a0d57958bc
Copying blob sha256:c6560e7248d7a1eab1c97b2b8d861dda677009bbb0bff7bcfdf59cb4043cdb07
Copying config sha256:a815ce10f90c5bda3770614c22651b916354912b6023e5f69ddcc8c5d2e35346
Writing manifest to image destination
Storing signatures
Generating dockerfile with builder image image-registry.openshift-image-registry.svc:5000/openshift/dotnet@sha256:73237ba08fab93d93075551a95b9e0047e765b0285d98d2fbffb9bbd691c4512
STEP 1: FROM image-registry.openshift-image-registry.svc:5000/openshift/dotnet@sha256:73237ba08fab93d93075551a95b9e0047e765b0285d98d2fbffb9bbd691c4512
STEP 2: LABEL "io.openshift.build.commit.ref"="HEAD" "io.openshift.build.commit.message"="Update for 2.1 release." "io.openshift.build.source-location"="https://github.com/redhat-developer/s2i-dotnetcore-ex.git" "io.openshift.build.source-context-dir"="/" "io.openshift.build.image"="image-registry.openshift-image-registry.svc:5000/openshift/dotnet@sha256:73237ba08fab93d93075551a95b9e0047e765b0285d98d2fbffb9bbd691c4512" "io.openshift.build.commit.author"="Tom Deseyn <tom.deseyn@gmail.com>" "io.openshift.build.commit.date"="Mon May 28 10:08:30 2018 +0200" "io.openshift.build.commit.id"="1ded43a7b3737ac31e31eee352134fd7c8c0d424"
STEP 3: ENV OPENSHIFT_BUILD_NAME="s-2-i-dotnetcore-ex-git-2" OPENSHIFT_BUILD_NAMESPACE="peopleapi" OPENSHIFT_BUILD_SOURCE="https://github.com/redhat-developer/s2i-dotnetcore-ex.git" OPENSHIFT_BUILD_COMMIT="1ded43a7b3737ac31e31eee352134fd7c8c0d424"
STEP 4: USER root
STEP 5: COPY upload/src /tmp/src
STEP 6: RUN chown -R 1001:0 /tmp/src
STEP 7: USER 1001
STEP 8: RUN /usr/libexec/s2i/assemble
---> Installing application source...
error: DOTNET_STARTUP_PROJECT has no project file
You can specify the startup project by adding an '.s2i/environment' file to the source repository.
The source repository contains the following projects:
- app/app.csproj
Update the '.s2i/environment' file to specify the project you want to publish, for example DOTNET_STARTUP_PROJECT=app/app.csproj.
subprocess exited with status 1
subprocess exited with status 1
error: build error: error building at STEP "RUN /usr/libexec/s2i/assemble": exit status 1

--

**Comment by omajid** [comment]
cc @aslicerh @tmds

**Comment by tmds** [comment]
```
error: DOTNET_STARTUP_PROJECT has no project file
You can specify the startup project by adding an '.s2i/environment' file to the source repository.
The source repository contains the following projects:

    app/app.csproj
    Update the '.s2i/environment' file to specify the project you want to publish, for example DOTNET_STARTUP_PROJECT=app/app.csproj.
```
You need to set the s2i context dir to `app`, or set the `DOTNET_STARTUP_PROJECT` to `app`.

**Comment by deenamanick** [comment]
DOTNET_STARTUP_PROJECT is working for Version 2.1 whereas it is failing for version 2.2. Do we need to check anything else? i managed to run the application using version 2.1 from the OpenShift catalog.

**Comment by tmds** [comment]
If you set `DOTNET_VERBOSITY` to some value (e.g. `normal`), the start of the build will print all environment variables. Try setting that and verify a. it prints out the environment variables, and b. `DOTNET_STARTUP_PROJECT` has the value you expect.

The tests we have in this repo fail if `DOTNET_STARTUP_PROJECT` doesn't work as expected. I will check manually using OpenShift 4.3 (today, or tomorrow).

**Comment by tmds** [comment]
> OpenShift 4.3 (today, or tomorrow).

I was hoping to run this using `crc`, but due to a bug, 4.3 is only supported from the next `crc` release.

@deenamanick have you had a chance to try the suggestions from my previous comment?
And does it behave the same if you use the exact commands that are mentioned in the docs: https://access.redhat.com/documentation/en-us/net_core/3.1/html/getting_started_guide/gs_dotnet_on_openshift#sample_app?

**Comment by deenamanick** [comment]
I cannot use the command to test since it is an Openshift Online. I doubt i will not be able to run oc command. let me try again and post the error for 2.3

**Comment by tmds** [comment]
> I was hoping to run this using crc, but due to a bug, 4.3 is only supported from the next crc release.

`crc` now supports running OpenShift 4.3.
I executed these commands on a new installation:
```
oc login -u developer -p developer https://api.crc.testing:6443
oc project
oc new-project test
./install-imagestreams.sh --os rhel7 --user <username> --password <password>
oc new-app dotnet:3.1~https://github.com/redhat-developer/s2i-dotnetcore-ex#dotnetcore-3.1 --context-dir=app
oc expose svc/s2i-dotnetcore-ex
```
The application deployed as expected.

(The `install-imagestreams.sh` script can be found in this repo.)

> I cannot use the command to test since it is an Openshift Online. I doubt i will not be able to run oc command. let me try again and post the error for 2.3

Also with OpenShift Online, you can use `oc`. You can find the login command via the top right menu.

**Comment by tmds** [comment]
Closing due to idle.
